### ISSUE TITLE
Race condition in PluginCredentialProvider results in failure to restore packages

### ISSUE THREAD

**Issue by zarenner** [NuGet/Home#1902]
A customer reported that they were encountering "unable to find version" errors restoring packages while using a credential provider.  An example of such error would be:

```
 WARNING: Unable to find version '1.0.0' of package 'TestPackage'.
 Unable to find version '1.0.0' of package 'TestPackage'.
```

We've only been able to reproduce the issue if the following is true:
- Restoring from multiple (in this case a mix of v2 and v3) package sources
- The credential provider (in this case CredentialProvider.VSS.exe") is copied alongside nuget.exe immediately before running nuget restore.  Something about the timing of a freshly copied executable seems to increase the likelihood of the race condition.

After investigation, we've determined that the bug is exposed when on line 152 of  https://github.com/NuGet/NuGet.Client/blob/dev/src/NuGet.Clients/NuGet.Credentials/PluginCredentialProvider.cs, cancellationToken.Register is called.  In these rare cases, the cancellation token's source has already been disposed, so the call to cancellationToken.Register throws an ObjectDisposedException.  This causes the CredentialProvider process to continue running (as PluginCredentialProvider.Execute throws/returns and is handled higher in the stack), yet the semaphore in CredentialService is released while the process is still running.

The CredentialProvider process then is executed again for a later request (but using the same instance of PluginCredentialProvider), and the ReadStdOut method receives the OutputDataReceived events for _both_ processes.  That is, _stdOut/responseJson contains two json objects, and we fail to deserialize the json.

There are a few things that need to be further investigated and fixed here:
1.  Why is the cancellation token source disposed?  I haven't yet determined when/where it's disposed, but it seems that something higher in the stack is disposing it earlier than it should.
2.  There is likely another bug here.  In normal operation, PluginCredentialProvider calls process.WaitForExit(Int32).  The [documentation](https://msdn.microsoft.com/en-us/library/ty0d8k56%28v=vs.110%29.aspx) for WaitForExit states that when using asynchronous event handlers, WaitforExit(Int32) calls must be followed with a call to WaitForExit() to ensure that asynchronous event handling is completed.  Currently, I think it's possible that we could start another CredentialProvider process after the first one exits, but before all output is received.
3.  _stdOut + _stdError should be made into local variables, and ReadStdOut + ReadStdError made into anonymous functions, to provide additional safety here.  There's no reason they need to be members.

**Comment by yishaigalatzer** [comment]
@zarenner I assume you guys are working on a PR for this?

**Comment by zarenner** [comment]
@yishaigalatzer Yes, we will put out a PR for the changes in PluginCredentialProvider (2 and 3 in the above list).  As far as why the cancellation token source is disposed early... I don't know enough yet.  If someone on your side has an idea, it would be helpful.

**Comment by jascam** [comment]
@yishaigalatzer For item 1:
The cancelation token is being generated in https://github.com/NuGet/NuGet.Client/blob/dev/src/NuGet.Core/NuGet.PackageManagement/PackageDownloader.cs at line 47.  The token is being passed to PluginCredentialProvider.cs via the GetDownloadResourceResultAsync call at line 75.  The dispose is happening when the using is exited.  Looking at the code this is probably line 87.  Unfortunately there is no wait before exiting for the other tasks to finish after canceling.

Changing 2 or 3 is secondary to blocking dispose of the token until we know the tasks have exited.  Without that defense against disposed tokens should be put anywhere the token is evaluated anywhere in the GetDownloadResourceResultAsync stack.

Our PR will include the change for item 1.  I think 2 and 3 are good changes but not necessarily for this issue if we change 1 do you want those as part of this PR or separate PR?

**Comment by yishaigalatzer** [comment]
Up to you, as long as these are small fixes that make reviewing reasonable you can go with one pr

**Comment by jascam** [comment]
PR ready for review https://github.com/NuGet/NuGet.Client/pull/196

**Comment by yishaigalatzer** [comment]
Fixed in 3.4
